cpp/src/strings/json/json_path.cu

```diff
@@ -0,0 +1,983 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/get_value.cuh>
+#include <cudf/detail/null_mask.hpp>
+#include <cudf/detail/utilities/cuda.cuh>
+#include <cudf/scalar/scalar.hpp>
+#include <cudf/scalar/scalar_factories.hpp>
+#include <cudf/strings/string_view.cuh>
+#include <cudf/strings/strings_column_view.hpp>
+#include <cudf/types.hpp>
+#include <cudf/utilities/bit.hpp>
+#include <cudf/utilities/error.hpp>
+
+#include <io/utilities/column_type_histogram.hpp>
+#include <io/utilities/parsing_utils.cuh>
+
+#include <rmm/device_uvector.hpp>
+#include <rmm/exec_policy.hpp>
+
+namespace cudf {
+namespace strings {
+namespace detail {
+
+namespace {
+
+// debug accessibility
+
+// change to "\n" and 1 to make output more readable
+#define DEBUG_NEWLINE
+constexpr int DEBUG_NEWLINE_LEN = 0;
+
+// temporary? spark doesn't strictly follow the JSONPath spec.
+// I think this probably could be a configurable enum to control
+// the kind of output you get and what features are supported.
+//
+// Current known differences:
+// - When returning a string value as a single element, Spark strips the quotes.
+//   standard:   "whee"
+//   spark:      whee
+//
+// - Spark only supports the wildcard operator when in a subscript, eg  [*]
+//   It does not handle .*
+//
+// Other, non-spark known differences:
+//
+// - In jsonpath_ng, name subscripts can use double quotes instead of the standard
+//   single quotes in the query string.
+//   standard:      $.thing['subscript']
+//   jsonpath_ng:   $.thing["subscript"]
+//
+//  Currently, this code only allows single-quotes but that could be expanded if necessary.
+//
+#define SPARK_BEHAVIORS
```

Comment:
This isn't intended to be kept, is it?

Reply:
Yes.  I think eventually we'll need to have a number of custom behaviors (not driven through a #define though).  But for now i wanted to keep this as a clear indicator of special behavior.

Reply:
I strongly dislike this precedent. Behaviors in the C++ library should not be defined in terms of a high-level interface. Specializing behavior is fine, but the definition and description of the specialization should not rely on references to Spark, Pandas, etc.

Reply:
What's the alternative?  I could remove the #define but we definitely need to clearly document the place that uses it, as it is a behavior that is absolutely unique to Spark.